restapi/src/main/java/io/stargate/web/docsapi/service/ReactiveDocumentService.java

```diff
@@ -138,4 +137,44 @@ public ReactiveDocumentService(
                       }));
         });
   }
+
+  public ObjectNode createJsonMap(DocumentDB db, List<RawDocument> docs, List<String> fields) {
+    ObjectNode docsResult = objectMapper.createObjectNode();
+
+    for (RawDocument doc : docs) {
+      // filter needed rows only
+      List<Row> rows = doc.rows();
+      if (!fields.isEmpty()) {
+        rows =
+            doc.rows().stream()
+                .filter(row -> fields.stream().anyMatch(field -> fieldMatchesRowPath(row, field)))
+                .collect(Collectors.toList());
+      }
+
+      // create document node and set to result
+      JsonNode node = jsonConverter.convertToJsonDoc(rows, false, db.treatBooleansAsNumeric());
+      docsResult.set(doc.id(), node);
+    }
+
+    return docsResult;
+  }
+
+  private boolean fieldMatchesRowPath(Row row, String field) {
+    String rowPath = getFieldPathFromRow(row, DocumentServiceUtils.maxFieldDepth(field));
+    return rowPath.startsWith(field)
+        && (rowPath.substring(field.length()).isEmpty()
+            || rowPath.substring(field.length()).startsWith("."));
+  }
+
+  private String getFieldPathFromRow(Row row, long maxDepth) {
+    List<String> path = new ArrayList<>();
+    for (int i = 0; i < maxDepth; i++) {
+      String value = row.getString("p" + i);
+      if (value == null || value.isEmpty()) {
+        break;
+      }
+      path.add(value);
+    }
+    return String.join(".", path);
```

Comment:
Each request has a few fields, but may have many rows. Could we refactor the filtering logic to use a list of individual field names (at each level) and compare them to `pNNN` values to avoid constructing transient `String`s?

Reply:
Sure we can, I already optimized here, but due to some reverts this was gone.. Good catch :+1: